pkg/resource/secret_reconciler.go

```diff
@@ -68,53 +68,85 @@ func NewSecretPropagatingReconciler(m manager.Manager) reconcile.Reconciler {
 		ctx, cancel := context.WithTimeout(context.Background(), secretReconcileTimeout)
 		defer cancel()
 
-		// The 'from' secret is also know as the 'propagating' secret.
-		from := &corev1.Secret{}
-		if err := client.Get(ctx, req.NamespacedName, from); err != nil {
-			// There's no propagation to be done if the secret we're propagating
-			// from does not exist. We assume we have a watch on that secret and
-			// will be queued if/when it is created. Otherwise we'll be requeued
-			// implicitly because we return an error.
-			return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetSecret)
-		}
-
 		// The 'to' secret is also known as the 'propagated' secret. We guard
 		// against abusers of the propagation process by requiring that both
 		// secrets consent to propagation by specifying each other's UID. We
 		// cannot know the UID of a secret that doesn't exist, so the propagated
 		// secret must be created outside of the propagation process.
 		to := &corev1.Secret{}
-		n := types.NamespacedName{
-			Namespace: from.GetAnnotations()[AnnotationKeyPropagateToNamespace],
-			Name:      from.GetAnnotations()[AnnotationKeyPropagateToName],
-		}
-		if err := client.Get(ctx, n, to); err != nil {
+		if err := client.Get(ctx, req.NamespacedName, to); err != nil {
 			// There's no propagation to be done if the secret we propagate to
 			// does not exist. We assume we have a watch on that secret and will
 			// be queued if/when it is created. Otherwise we'll be requeued
 			// implicitly because we return an error.
 			return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetSecret)
 		}
 
-		if from.GetAnnotations()[AnnotationKeyPropagateToUID] != string(to.GetUID()) {
-			// The propagating secret expected a different propagated secret. We
+		toAnnotations := to.GetAnnotations()
+		var fromUID string
+		var fromName string
+		var fromNamespace string
+
+		for key, val := range toAnnotations {
+			// If this is a 'from' secret we do not want to reconcile and will
+			// instead defer to the enqueued request for the 'to' secret
+			if strings.HasPrefix(key, AnnotationKeyPropagateTo) {
```

Comment:
Is there a better way to do this?

Reply:
Do we need to consider the case in which a secret is both propagated and a propagator? I think that would work (by accident) in the contemporary implementation, though you'd have to manually write annotations for it to ever happen.

Reply:
Hmm I was thinking about this and I believe you are correct that this could only happen with manual annotations. So then I was thinking of whether this is something we should allow or not. The purpose of the secret reconciler is to propagate data from one secret to another when they both consent to the process. Technically a chain of secrets that do this should still be able to be handled by this reconciler as long as each consents to propagation with the next and preceding secrets. While I am not sure of a use-case for this, I am leaning towards allowing it in the context of this reconciler. What do you think?

Reply:
Yeah, given it doesn't seem to a a risk I'd lean toward allowing it, especially if it simplifies the code.